### PR TITLE
Backend enhancements

### PR DESCRIPTION
**Pull request by c0derMo**
This PR fixes and enhances a couple minor things in the backend server, namely:
- Added sessions and middleware to enforce user permissions
- Added custom error types to allow for quicker throwing of errors with codes
- Port, Hostname, Database can be configured through environment variables and `.env`-files
- Backend server serves the built frontend, if thats available
- Support for HTTPS

### CODE REVIEW

File: backend/loernwerkServer.ts [hunk 1]
```diff
@@ -28,10 +35,27 @@ class loernwerkServer {
...
 
+        // Setting up cookies
+        app.use(
+            session({
+                name: 'loernwerk.session',
+                resave: false,
+                saveUninitialized: false,
+                secret: randomBytes(64).toString('hex'),
+                cookie: {
+                    maxAge: 1000 * 60 * 60 * 24,
```
Comment (Kr0nox) [inline backend/loernwerkServer.ts:47]: add a comment as to what time frame this is

File: backend/loernwerkServer.ts [hunk 2]
```diff
@@ -28,10 +35,27 @@ class loernwerkServer {
...
+                },
+                store: new (MemoryStore(session))({
+                    checkPeriod: 1000 * 60 * 60,
```
Comment (Kr0nox) [inline backend/loernwerkServer.ts:51]: add a comment as to what time frame this is

**Review by a-bumblebee — APPROVED** [review]

**Review by Kr0nox — CHANGES_REQUESTED** [review]
2 things that should ebe specified in more detail.

One addtional question: When a https server is server, does a http server always get started as well?

File: backend/loernwerkServer.ts (hunk 1 above)
Comment (c0derMo) [reply backend/loernwerkServer.ts:47]: true, 24h is default session length because of teachers workflows

File: backend/loernwerkServer.ts (hunk 2 above)
Comment (c0derMo) [reply backend/loernwerkServer.ts:51]: purging expired sessions every hour

**Comment by c0derMo** [comment]
> When a https server is server, does a http server always get started as well?

Currently yes, HTTP is always available, HTTPS only if the certificate & key are supplied. We could add an option to disable HTTP if you think that's needed

**Comment by Kr0nox** [comment]
> We could add an option to disable HTTP if you think that's needed

What are the benefits of having it always disabled and what are the disadvantages if it is deisabled?

**Comment by c0derMo** [comment]
Not sure, incorrect setup and expired certificates might make accessing the platform more complicated if HTTP is disabled, but I don't think there's any major disadvantage, just thought it was the "usual way" platforms are always reachable through both HTTP and HTTPS

**Comment by Kr0nox** [comment]
I would have suggested given a system admin the option to start the service without http.
Use case might be if the communication needs to be encrypted and therefor should not run over http

**Comment by c0derMo** [comment]
Added an environment variable which, if set, disables the HTTP server

**Review by Kr0nox — APPROVED** [review]

**Pull request merged**